sdk/azidentity/azure_cli_credential.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+type AzureCLITokenProvider interface {
+	GetCLIToken(ctx context.Context, resource string) ([]byte, error)
+}
+
+type azureCLITokenProvider struct{}
+
+// AzureCLICredentialOptions contains options used to configure the AzureCLICredential
+type AzureCLICredentialOptions struct {
+	TokenProvider AzureCLITokenProvider
+}
+
+// AzureCLICredential enables authentication to Azure Active Directory using the Azure CLI command "az account get-access-token".
+type AzureCLICredential struct {
+	tokenProvider AzureCLITokenProvider
+}
+
+// TODO: do we want an options bag for this credential that includes a developer specified Azure CLI path or only specify with env var?
+// NewAzureCLICredential constructs a new AzureCLICredential with the details needed to authenticate against Azure Active Directory
+// options: configure the management of the requests sent to Azure Active Directory.
+func NewAzureCLICredential(options *AzureCLICredentialOptions) (*AzureCLICredential, error) {
+	if options != nil {
+		return &AzureCLICredential{
+			tokenProvider: options.TokenProvider,
+		}, nil
+	}
+	return &AzureCLICredential{
+		tokenProvider: &azureCLITokenProvider{},
+	}, nil
+}
+
+// GetToken obtains a token from Azure Active Directory, using the Azure CLI command to authenticate.
+// ctx: Context used to control the request lifetime.
+// opts: TokenRequestOptions contains the list of scopes for which the token will have access.
+// Returns an AccessToken which can be used to authenticate service client calls.
+func (c *AzureCLICredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
+	// The following code will remove the /.default suffix from the scope passed into the method since AzureCLI expect a resource string instead of a scope string
+	if strings.HasSuffix(opts.Scopes[0], defaultSuffix) {
```

Comment:
This seems a bit wonky, we shouldn't be repurposing `Scopes` to really mean an Azure resource.  If scopes aren't applicable to CLI credentials we should simply ignore them and pass the resource via other means.

Reply:
We also do this in MSI credential, agree it's not great, but if we want a specific resources var then we would need to add it in TokenRequestOptions

Reply:
I would prefer we do that then as the intent is (somewhat) clear.  We'll have to ensure to properly document the cases where `Scopes` and `Resource` are required and gracefully fail when they're missing.

Reply:
So in order to do this I think we would probably have to have a separate `resources` field on TokenRequestOptions in order to differentiate between scopes and resources. Then when we call the AuthenticationPolicy pass in any scopes or resource strings we might have. I'll make this proposed change in a separate PR.